### DISCUSSION TITLE
Scheduled job depends on asset - how to ensure materialization at runtime

### DISCUSSION THREAD

**Discussion by lyschoening** [dagster-io/dagster#17561] [Q&A]
Hello,

I am following the dagster-email-report example and would like to configure the job (https://github.com/dagster-io/dagster-email-report/blob/main/user_report/__init__.py#L16) so that it ensures that the asset it uses is materialized as needed. 

Since the job converts the asset to a source asset, I can't count on auto-materialization. I've seen suggestions for adding an asset sensor, but I am unsure whether that would work for me as the asset might be materialized more often than the job is scheduled.

**Comment by sryza** [comment]
> so that it ensures that the asset it uses is materialized as needed.

Are you able to explain what you mean by "materialized as needed"? Is the behavior that you're looking for that you want the  job to run once per month (or some other cadence), but only if the asset has been updated since the last run?

**Reply by lyschoening** [reply]
I want the job to run once a month and I want to materialize the asset if it hasn't already at the time it runs. For example dagster-email-report has a scheduled job that requires an asset, https://github.com/dagster-io/dagster-email-report/blob/main/user_report/assets.py#L218C17-L218C31, so the job will fail if the asset (or a partition) is not materialized.

As mentioned before, I am unsure how to use an asset sensor, i.e. where the asset materialization triggers the job, because the job has the side effect of sending an email. I would not want to send another email just because someone decides to re-materialize the asset in the UI, for example.

**Reply by sryza** [reply]
This isn't a full answer to your question, but my general recommendation here would be:
- Model the email report as an asset itself, rather than an op-based job. While this stretches the definition of asset a bit, it's mostly harmless and will allow you launch a single run that materializes the asset and sends the email report.
- Write a regular sensor with whatever logic you want, that selects either one or both assets for execution, depending on the circumstances.
